### PR TITLE
reset: open file for write when file exists and skipworktree bit is on

### PR DESCRIPTION
**Pull request by kewillford**
When running a virtual file system and an entry with the skipworktree
bit on has only been read, meaning it has not yet been added to the
modified paths list, and the reset is going to remove it from
the index, the file needs to be opened for write so that the path
will be added to the modified paths and git will pick it up as an
untracked file.

### CODE REVIEW

File: builtin/reset.c [hunk 1]
```diff
@@ -162,7 +163,12 @@ static void update_index_from_diff(struct diff_queue_struct *q,
 					die(_("make_cache_entry failed for path '%s'"),
 						two->path);
 
-				checkout_entry(ceBefore, &state, NULL);
+				if (file_exists(two->path)) {
```
Comment (benpeart) [inline builtin/reset.c]: It looks like this logic is very specific to GVFS and our knowledge of how it behaves.  As such, it should be wrapped with the appropriate gvfs_config_is_set() test.

The rest should probably be submitted upstream as it improves a boundary case that exists for regular sparse checkout repos.

Comment (wilbaker) [reply builtin/reset.c]: As discussed offline, rather than opening the file for write can we try clearing the file's skip-worktree bit?  VFS for Git should pick that change up and automatically add the path to ModifiedPaths.dat.
